src/Polysemy/Internal/Combinators.hs

```diff
@@ -160,14 +159,14 @@ lazilyStateful f = interpretInLazyStateT $ \e -> LS.StateT $ fmap swap . f e
 -- See the notes on 'Tactical' for how to use this function.
 reinterpretH
     :: forall e1 e2 r a
-     . (∀ m x. e1 m x -> Tactical e1 m (e2 ': r) x)
+     . (∀ m x. Monad m => e1 m x -> Tactical e1 m (e2 ': r) x)
```

Comment:
This or `∀ r0 x. e1 (Sem r0) x -> Tactical e1 (Sem r0) (e2 ': r) x`? Yours certainly is prettier, in that it exposes less potentially confusing details, but I'm worried about two things:

Performance: This is the big one. I would absolutely hate it if GHC would pass `Sem r0`'s monad dictionary into this function instead of using it directly, and I believe that, unfortunately, that's exactly what it will do. Maybe it will inline and specialize upon the surface call, thanks to `loop-breaker`, but I believe that, no matter what, it will do the dictionary-passing on recursive calls. Specifying `Sem r0` is a sure way to prevent this kind of dictionary-passing from happening.

Flexibility: Even without knowing `r0`, there are some stuff you can do with an `Sem r0` by virtue of it being a free monad, like splitting it into layers of actions, which you could then muck around with. This is very contrived argument however, and I can't imagine it actually mattering. Also, exposing `m ~ Sem r0` could prove necessary to solve #314, but that's speculation.

Reply:
I have fixed all instances of this to use the approach you suggested.